tracer/src/Datadog.Trace/Configuration/ImmutableTracerSettings.cs

```diff
@@ -0,0 +1,333 @@
+        /// Gets a value indicating custom sampling rules.
+        /// </summary>
+        /// <seealso cref="ConfigurationKeys.CustomSamplingRules"/>
+        public string CustomSamplingRules { get; }
+
+        /// <summary>
+        /// Gets a value indicating a global rate for sampling.
+        /// </summary>
+        /// <seealso cref="ConfigurationKeys.GlobalSamplingRate"/>
+        public double? GlobalSamplingRate { get; }
+
+        /// <summary>
+        /// Gets a collection of <see cref="Integrations"/> keyed by integration name.
+        /// </summary>
+        public ImmutableIntegrationSettingsCollection Integrations { get; }
+
+        /// <summary>
+        /// Gets the global tags, which are applied to all <see cref="Span"/>s.
+        /// </summary>
+        public IDictionary<string, string> GlobalTags { get; }
+
+        /// <summary>
+        /// Gets the map of header keys to tag names, which are applied to the root <see cref="Span"/> of incoming requests.
+        /// </summary>
+        public IDictionary<string, string> HeaderTags { get; }
+
+        /// <summary>
+        /// Gets the port where the DogStatsd server is listening for connections.
+        /// Default is <c>8125</c>.
+        /// </summary>
+        /// <seealso cref="ConfigurationKeys.DogStatsdPort"/>
+        public int DogStatsdPort { get; }
+
+        /// <summary>
+        /// Gets a value indicating whether internal metrics
+        /// are enabled and sent to DogStatsd.
+        /// </summary>
+        public bool TracerMetricsEnabled { get; }
+
+        /// <summary>
+        /// Gets a value indicating whether runtime metrics
+        /// are enabled and sent to DogStatsd.
+        /// </summary>
+        public bool RuntimeMetricsEnabled { get; }
+
+        /// <summary>
+        /// Gets a value indicating whether partial flush is enabled
+        /// </summary>
+        public bool PartialFlushEnabled { get; }
+
+        /// <summary>
+        /// Gets the minimum number of closed spans in a trace before it's partially flushed
+        /// </summary>
+        public int PartialFlushMinSpans { get; }
+
+        /// <summary>
+        /// Gets a value indicating whether a span context should be created on exiting a successful Kafka
+        /// Consumer.Consume() call, and closed on entering Consumer.Consume().
+        /// </summary>
+        /// <seealso cref="ConfigurationKeys.KafkaCreateConsumerScopeEnabled"/>
+        public bool KafkaCreateConsumerScopeEnabled { get; }
+
+        /// <summary>
+        /// Gets a value indicating whether the diagnostic log at startup is enabled
+        /// </summary>
+        public bool StartupDiagnosticLogEnabled { get; }
+
+        /// <summary>
+        /// Gets the comma separated list of url patterns to skip tracing.
+        /// </summary>
+        /// <seealso cref="ConfigurationKeys.HttpClientExcludedUrlSubstrings"/>
+        internal string[] HttpClientExcludedUrlSubstrings { get; }
```

Comment:
IReadOnlyCollection?

Reply:
I decided not to convert these to interfaces as they're already _kinda_ immutable (I know, I know, not really), and it didn't seem worth taking the potential performance hit for the interface dispatch?

Reply:
Normally I'd say interface dispatch isn't anything to be afraid of, but I guess if these are on one of the tracer's hot paths then it could make a difference.